zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs

```diff
@@ -0,0 +1,358 @@
+        };
+
+        if is_root {
+            code_b.append(&bytecode! {
+                PUSH32(return_data_size)
+                PUSH32(return_data_offset)
+                RETURN
+                STOP
+            });
+        } else {
+            code_b.append(&bytecode! {
+                PUSH32(size) // size
+                PUSH32(offset) // data offset
+                PUSH32(dest_offset) // memory offset
+                RETURNDATACOPY
+                // end for internal
+                PUSH32(return_data_size)
+                PUSH32(return_data_offset)
+                RETURN
+                STOP
+            });
+        }
+
+        // code A calls code B.
+        let mut code_a = bytecode! {
+            // call ADDR_B.
+            PUSH32(return_data_size) // retLength
+            PUSH32(return_data_offset) // retOffset
+            PUSH1(0x00) // argsLength
+            PUSH1(0x00) // argsOffset
+            PUSH1(0x00) // value
+            PUSH32(addr_b.to_word()) // addr
+            PUSH32(0x1_0000) // gas
+            CALL
+        };
+
+        if is_root {
+            code_a.append(&bytecode! {
+                PUSH32(size) // size
+                PUSH32(offset) // offset
+                PUSH32(dest_offset) // dest_offset
+                RETURNDATACOPY
+                STOP
+            });
+        } else {
+            code_a.append(&bytecode! {
+                PUSH32(return_data_size)
+                PUSH32(return_data_offset)
+                RETURN
+            });
+        }
+
+        let ctx = TestContext::<3, 1>::new(
+            None,
+            |accs| {
+                accs[0].address(addr_a).code(code_a);
+                accs[1].address(addr_b).code(code_b);
+                accs[2]
+                    .address(mock::MOCK_ACCOUNTS[2])
+                    .balance(Word::from(1u64 << 30));
+            },
+            |mut txs, accs| {
+                txs[0].to(accs[0].address).from(accs[2].address);
+            },
+            |block, _tx| block,
+        )
+        .unwrap();
+
+        assert_eq!(
+            run_test_circuits_with_params(
+                ctx,
+                None,
+                CircuitsParams {
+                    max_rws: 2048,
+                    ..Default::default()
+                }
+            ),
+            Ok(())
+        );
+    }
+
+    // test root & internal calls
+    #[test]
+    fn returndatacopy_out_of_bound_error() {
+        // test root call cases: `end` exceed return data size
+        test_ok_internal(0x00, 0x10, 0x20, 0x10, 0x10, true);
+        // test internal call case: `end` exceed return data size
+        test_ok_internal(0x00, 0x10, 0x20, 0x10, 0x10, false);
+        // test data offset overflow
+        test_ok_internal(0x00, 0x10, 0x20, u128::from(u64::MAX) + 1, 0x10, true);
+        // test end = data offset + length(size) overflow
+        test_ok_internal(0x00, 0x10, 0x20, u128::from(u64::MAX) - 5, 0x10, true);
```

Comment:
To test end overflow, we should make the offset smaller.
```suggestion
        test_ok_internal(0x00, 0x10, 0x20, 0x1, 0x10, true);
```

Reply:
updated in [f3bc7ac](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1017/commits/f3bc7ac6898a06c6ef724f12875b74410bb0a81d) thanks!